packages/components/src/tools-panel/test/index.js

```diff
@@ -348,6 +349,237 @@ describe( 'ToolsPanel', () => {
+					) }
+				</ToolsPanel>
+			);
+
+			const { rerender } = render( <TestPanel /> );
+
+			// The conditional control should not yet be rendered.
+			let conditionalItem = screen.queryByText( 'Conditional control' );
+			expect( conditionalItem ).not.toBeInTheDocument();
+
+			// The conditional control should not yet appear in the default controls
+			// menu group.
+			openDropdownMenu();
+			let menuGroups = screen.getAllByRole( 'group' );
+			let defaultItem = within( menuGroups[ 0 ] ).queryByText(
+				'Conditional'
+			);
+			expect( defaultItem ).not.toBeInTheDocument();
+
+			// Simulate the main control having a value set which will now
+			// render the new default control into the ToolsPanel.
+			altControlProps.attributes.value = true;
+
+			rerender( <TestPanel /> );
+
+			// The conditional control should now be rendered and included in
+			// the panel's menu.
+			conditionalItem = screen.getByText( 'Conditional control' );
+			expect( conditionalItem ).toBeInTheDocument();
+
+			// The conditional control should now appear in the default controls
+			// menu group.
+			menuGroups = screen.getAllByRole( 'group' );
+
+			// The new default control item for the Conditional control should
+			// be within the first menu group.
+			defaultItem = within( menuGroups[ 0 ] ).getByText( 'Conditional' );
+			expect( defaultItem ).toBeInTheDocument();
+		} );
+	} );
+
+	describe( 'registration of panel items', () => {
+		it( 'should register and deregister items when switching between panels', () => {
+			// This test adds a `panelId` prop to two panels, to simulate
+			// the registration and de-registration of panelItems when
+			// switching between panels and test for any issues.
+			//
+			// When you switch from panel A to panel B, only panel A items
+			// should be deregistered, and only panel B items should be registered.
+			//
+			// See: https://github.com/WordPress/gutenberg/pull/36588
+
+			const noop = () => undefined;
+			const context = {
+				panelId: '1234',
+				menuItems: {
+					default: {},
+					optional: { [ altControlProps.label ]: true },
+				},
+				hasMenuItems: false,
+				isResetting: false,
+				shouldRenderPlaceholderItems: false,
+				registerPanelItem: jest.fn(),
+				deregisterPanelItem: jest.fn(),
+				flagItemCustomization: noop,
+				areAllOptionalControlsHidden: true,
+			};
+
+			const TestPanel = () => (
+				<ToolsPanelContext.Provider value={ context }>
+					<ToolsPanelItem { ...altControlProps } panelId="1234">
+						<div>Item</div>
+					</ToolsPanelItem>
+				</ToolsPanelContext.Provider>
+			);
+
+			// On the initial render of the panel, the ToolsPanelItem should
+			// be registered.
+			const { rerender } = render( <TestPanel /> );
+
+			expect( context.registerPanelItem ).toHaveBeenCalledWith(
+				expect.objectContaining( {
+					label: altControlProps.label,
+					panelId: '1234',
+				} )
+			);
+			expect( context.deregisterPanelItem ).toHaveBeenCalledTimes( 0 );
```

Comment:
Tiny nitpick:

It looks like the prevailing approach to this is to use `.not.toHaveBeenCalled()`. Obviously, it makes no functional difference however it is probably a touch more readable.

```suggestion
			expect( context.deregisterPanelItem ).not.toHaveBeenCalled();
```